File: solutions/csharp/concertticketbookingsystem/ConcertTicketBookingSystem.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace ConcertBookingSystem
{
    public class ConcertTicketBookingSystem
    {
        private static ConcertTicketBookingSystem _instance;
        private readonly ConcurrentDictionary<string, Concert> concerts;
        private readonly ConcurrentDictionary<string, Booking> bookings;
        private readonly object lockObject = new object();

        private ConcertTicketBookingSystem()
        {
            concerts = new ConcurrentDictionary<string, Concert>();
            bookings = new ConcurrentDictionary<string, Booking>();
        }

        public static ConcertTicketBookingSystem Instance
        {
            get
            {
                if (_instance == null)
                {
                    _instance = new ConcertTicketBookingSystem();
                }
                return _instance;
            }
        }

        public void AddConcert(Concert concert)
        {
            concerts.TryAdd(concert.Id, concert);
        }

        public Concert GetConcert(string concertId)
        {
            concerts.TryGetValue(concertId, out var concert);
            return concert;
        }

        public List<Concert> SearchConcerts(string artist, string venue, DateTime dateTime)
        {
            return concerts.Values.Where(concert =>
                concert.Artist.Equals(artist, StringComparison.OrdinalIgnoreCase) &&
                concert.Venue.Equals(venue, StringComparison.OrdinalIgnoreCase) &&
                concert.DateTime == dateTime).ToList();
        }

        public Booking BookTickets(User user, Concert concert, List<Seat> seats)
        {
            lock (lockObject)
            {
                foreach (var seat in seats)
                {
                    if (seat.Status != SeatStatus.AVAILABLE)
                    {
                        throw new SeatNotAvailableException($"Seat {seat.SeatNumber} is not available.");
                    }
                }

                seats.ForEach(seat => seat.Book());

                string bookingId = GenerateBookingId();
                var booking = new Booking(bookingId, user, concert, seats);
                bookings.TryAdd(bookingId, booking);

                ProcessPayment(booking);
                booking.ConfirmBooking();

                Console.WriteLine($"Booking {booking.Id} - {booking.Seats.Count} seats booked");

                return booking;
            }
        }

        public void CancelBooking(string bookingId)
        {
            bookings.TryRemove(bookingId, out var booking);
            booking?.CancelBooking();
        }

        private void ProcessPayment(Booking booking)
        {
            // Simulate payment processing logic here
        }

        private string GenerateBookingId()
        {
            return "BKG" + Guid.NewGuid();
        }
    }
}

File: solutions/python/concertticketbookingsystem/concertticketbookingsystem.py
from datetime import datetime
from threading import Lock
from typing import List
from concert import Concert
from user import User
from seat import Seat, SeatNotAvailableException
from booking import Booking
from enums import SeatStatus
import uuid
class ConcertTicketBookingSystem:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance.concerts = {}
            cls._instance.bookings = {}
            cls._instance._lock = Lock()
        return cls._instance
    
    def addConcert(self, concert: Concert):
        self.concerts[concert.id] = concert

    def getConcert(self, concertId: str) -> "Concert":
        if concertId in self.concerts:
            return self.concerts[concertId]
        print(f"No concert with concertId {concertId} found")
    
    def searchConcerts(self, artist: str, venue: str, dateTime: datetime) -> List["Concert"]:
        return [ concert for _, concert in self.concerts.items() \
            if concert.getArtist() == artist \
            and concert.getVenue() == venue \
            and concert.getDateTime() == dateTime ]
    
    def bookTickets(self, user: User, concert: Concert, seats: List["Seat"]) -> "Booking":
        with self._lock:
            # ensure all the seats are available
            for seat in seats:
                if seat.seatStatus != SeatStatus.AVAILABLE:
                    raise SeatNotAvailableException("Seat not available")
            for seat in seats:
                # Reserve all seats
                seat.book()
            
            bookingId = self._generateId(user, concert)
            booking = Booking(bookingId, user, concert, seats)

            self.bookings[bookingId] = booking
            self._processPayment(booking)
            booking.confirmBooking()
            print(f"Booking {bookingId} confirmed")
            return booking
    
    def cancelBooking(self, bookingId: str):
        with self._lock:
            if bookingId in self.bookings:
                booking = self.bookings.pop(bookingId)
                booking.cancelBooking()

    def _processPayment(self, booking: Booking) -> bool:
        price = booking.getTotalPrice()
        print(f"Processed payment for amount {price}")
        return True
    
    def _generateId(self, user: User, concert: Concert) -> str:
        return f"BKG{user.id}-{concert.id}-{uuid.uuid4()}"